[SPARK-27324][DOC][CORE] Document configurations related to executor metrics and modify a configuration.

wypoon opened the pull request:
### What changes were proposed in this pull request?

Add a section to the Configuration page to document configurations for executor metrics.
At the same time, rename spark.eventLog.logStageExecutorProcessTreeMetrics.enabled to spark.executor.processTreeMetrics.enabled and make it independent of spark.eventLog.logStageExecutorMetrics.enabled.

### Why are the changes needed?

Executor metrics are new in Spark 3.0. They lack documentation.
Memory metrics as a whole are always collected, but the ones obtained from the process tree have to be optionally enabled. Making this depend on a single configuration makes for more intuitive behavior. Given this, the configuration property is renamed to better reflect its meaning.

### Does this PR introduce any user-facing change?

Yes, only in that the configurations are all new to 3.0.

### How was this patch tested?

Not necessary.

wypoon committed:
[SPARK-27324][DOC] Document configurations related to executor metrics.

SparkQA commented:
**[Test build #117264 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117264/testReport)** for PR 27329 at commit [`7901d1c`](https://github.com/apache/spark/commit/7901d1c18845050682744efdffc2a4654d182341).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22025/
Test PASSed.

wypoon commented:
@rezasafi I'd like to propose renaming `spark.eventLog.logStageExecutorProcessTreeMetrics.enabled` to say, `spark.executor.processTreeMetrics.enabled`. Unlike `spark.eventLog.logStageExecutorMetrics.enabled`, which determines whether to **log** executor metrics (which are always collected), this property determines whether to collect the process tree metrics in the first place. As such, I think the current name is slightly misleading and could be improved. Since all these configurations are new in 3.0, I'd like to do the renaming now, before 3.0 is released. What do you think?
@squito ?

SparkQA commented:
**[Test build #117264 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117264/testReport)** for PR 27329 at commit [`7901d1c`](https://github.com/apache/spark/commit/7901d1c18845050682744efdffc2a4654d182341).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117264/
Test PASSed.

rezasafi commented:
@wypoon processtree metrics will also be written to eventlogs like other executor metrics and they will be gathered only if spark.eventLog.logStageExecutorMetrics.enabled is set to true. I think the current name which is "spark.eventLog.logStageExecutorProcessTreeMetrics.enabled" is self explanatory about this behavior. So If you want to change the name it makes more sense to change "spark.eventLog.logStageExecutorMetrics.enabled" as well. OtherwiseI think the current name is better to avoid user mis understanding.

wypoon committed:
[SPARK-27324] Rename spark.eventLog.logStageExecutorProcessTreeMetrics.enabled

... to spark.executor.processTreeMetrics.enabled.

SparkQA commented:
**[Test build #117331 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117331/testReport)** for PR 27329 at commit [`745d2d0`](https://github.com/apache/spark/commit/745d2d0925e0bc905734e94624b3fb206e398f9a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22091/
Test PASSed.

wypoon commented:
> @wypoon processtree metrics will also be written to eventlogs like other executor metrics and they will be gathered only if spark.eventLog.logStageExecutorMetrics.enabled is set to true. I think the current name which is "spark.eventLog.logStageExecutorProcessTreeMetrics.enabled" is self explanatory about this behavior. So If you want to change the name it makes more sense to change "spark.eventLog.logStageExecutorMetrics.enabled" as well. OtherwiseI think the current name is better to avoid user mis understanding.

It is not self-evident that spark.eventLog.logStageExecutorProcessTreeMetrics.enabled being true is only effective if spark.eventLog.logStageExecutorMetrics.enabled is true as well. This is one reason why we need documentation and I document this fact in this PR.
As I already pointed out, spark.eventLog.logStageExecutorMetrics.enabled is about whether to log metrics - the metrics (in general) are always collected - so that name is appropriate. spark.eventLog.logStageExecutorProcessTreeMetrics.enabled is about whether to collect a particular class of metrics in the first place. If they are not collected, they are logged as 0 (but still logged), if spark.eventLog.logStageExecutorMetrics.enabled is true.
This is why I propose the rename.

rezasafi commented:
I think when the name of the two options have the same prefix naturally it is more clear that the two metrics have some sort of a relation. I really not sure how much benefit a rename will have. 
Also your jira is about documentation and it seems weird to do a rename under that, however it isn't a weirdest thing that I have seen in this project!

SparkQA commented:
**[Test build #117331 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117331/testReport)** for PR 27329 at commit [`745d2d0`](https://github.com/apache/spark/commit/745d2d0925e0bc905734e94624b3fb206e398f9a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117331/
Test PASSed.

srowen commented:
No strong opinion on this here. 
If the metric being remove was never present in 2.x then it's OK, otherwise I think we'd have to deprecate, etc.

srowen reviewed (commented):
yes the change is straightforward, but what about `spark.eventLog.logStageExecutorMetrics.enabled`? Does it need to change too or is that pretty different?

wypoon commented:
> yes the change is straightforward, but what about `spark.eventLog.logStageExecutorMetrics.enabled`? Does it need to change too or is that pretty different?

That property does relate directly to the event log; it determines if aggregated metrics are written to the log. I did not feel the need to change that.

dongjoon-hyun commented:
@wypoon Could you update the PR title since this PR touches the code?

wypoon commented:
> @wypoon Could you update the PR title since this PR touches the code?

Done.

squito commented:
from a quick look, I think this makes sense (including the rename), but I need sometime to look back through the code and see where exactly these properties are touched to make sure.  It does seem like the property has more to do with whether executors collect the metrics.  Writing to the event log does naturally follow, btu you also get the metrics in any listener.

Also cc @LucaCanali who might have some thoughts here.

LucaCanali commented on docs/configuration.md:
```diff
@@ -1624,6 +1624,42 @@ Apart from these, the following properties are also available, and may be useful
+    How often to collect executor metrics.
+    <br />
+    If 0, the polling is done on executor heartbeats (thus at the heartbeat interval).
+    If positive, the polling is done at this interval.
```
I propose to add that the polling interval is expressed in milliseconds

wypoon replied:
Added unit.

LucaCanali commented on docs/configuration.md:
```diff
@@ -1624,6 +1624,42 @@ Apart from these, the following properties are also available, and may be useful
+  <td>
+    How often to collect executor metrics.
+    <br />
+    If 0, the polling is done on executor heartbeats (thus at the heartbeat interval).
```
We could explicitly reference the parameter spark.executor.heartbeatInterval

wypoon replied:
Added reference.

LucaCanali commented:
I think it is a good idea to add info on executor metrics configuration to `docs/configuration.md` as proposed in this PR.

The rationale discussed above for renaming `spark.eventLog.logStageExecutorProcessTreeMetrics.enabled` to `spark.executor.processTreeMetrics.enabled` seems valid to me. 

Additional comments:  
- This discussion actually exposes the strange relationship that the configuration `spark.eventLog.logStageExecutorProcessTreeMetrics.enabled` (here renamed to `spark.executor.processTreeMetrics.enabled`) has with `spark.eventLog.logStageExecutorMetrics.enabled`. Maybe there should not be the dependency that process tree metrics are collecteded if both `spark.eventLog.logStageExecutorMetrics.enabled` and `spark.executor.processTreeMetrics.enabled` are set to true, but we could change that and just use the parameter `spark.executor.processTreeMetrics.enabled` for this scope?

- Related to this discussion, I have also noticed that the description of `spark.eventLog.logStageExecutorMetrics.enabled` in `docs/monitoring.md` needs to be clarified, as it currently states "Metrics `peakExecutorMetrics.*` are only enabled if `spark.eventLog.logStageExecutorMetrics.enabled` is true".   
A change to that text could be added to this PR, following the parameter description in `configurations.md`?

wypoon commented:
@LucaCanali thanks for your comments.
At this point, I do not wish to change the behavior. Renaming a configuration property is one thing; it is a simple change. Changing the effect of a property is another thing, and requires testing. You can understand if I don't wish to take that on.
On your last point, I don't know if the statement in question is inaccurate (as I haven't examined that type of metrics). What I can clarify is that the executor always polls memory metrics (either in the heartbeat thread or in a separate thread) and sends them in its heartbeat. In the `EventLoggingListener`, there is logic to aggregate these memory metrics and log per-stage per-executor peaks on stage completion, but only if `spark.eventLog.logStageExecutorMetrics.enabled` is true (otherwise executor metrics update events are ignored).
Just to confirm, whenever the `ExecutorMetricsPoller` is asked to `poll`, then you update your `ExecutorMetricsSource` with the latest snapshot of the memory metrics, correct? In this case, this happens irrespective of  `spark.eventLog.logStageExecutorMetrics.enabled`. If this is correct, that the statement you refer to in `docs/monitoring.md` is inaccurate and should be corrected.

wypoon commented:
@LucaCanali thinking about it a bit more, if my understanding is correct, and memory metrics (except for process tree metrics) are always collected for the Dropwizard metrics, then I understand your wanting the behavior change you suggest (not have `spark.executor.processTreeMetrics.enabled` depend on `spark.eventLog.logStageExecutorMetrics.enabled`).

squito commented:
ok, yes I also see @rezasafi 's point now.  The procfs metrics are only gathered if event log stage executor metrics are turned on:

https://github.com/apache/spark/blob/f59685acaa3e9c227f14fe4d8f9e94a1ac664b05/core/src/main/scala/org/apache/spark/executor/ProcfsMetricsGetter.scala#L61-L65

Which makes the current name more reasonable.  But that is strange behavior, particularly with https://issues.apache.org/jira/browse/SPARK-27189 -- I should have thought to change that at the same time.

@wypoon I do think that there should also be behavior change as part of this, to remove that check in procfsmetricsgetter.

LucaCanali commented:
As far as I understand, also tested by running a few tests, the current situation is:  
As you wrote, `ExecutorMetricsSource` reports executor memory metrics values from the latest snapshot of the metrics as sent by the `ExecutorMetricsPoller`. These are not peak values of the metrics but rather represent the time evolution (for monitoring pourposes, handled via the dropwizard library, needs to be configured with the relevant parameters).  
Peak values of the executor memory metrics are reported by the Web UI under /api/v1/applications/.../executors for each executor. This information is made available irrespective of the value of `spark.eventLog.logStageExecutorMetrics.enabled`  
`SparkListenerStageExecutorMetrics` are logged to the eventlog with per-stage computed peak values of the executor metrics only if `spark.executor.processTreeMetrics.enabled` is set to true, as you wrote.

wypoon commented:
Thanks @LucaCanali and @squito. I will change the behavior in `ProcfsMetricsGetter` so collection of proc fs metrics do not depend on `spark.eventLog.logStageExecutorMetrics.enabled`, only on (the renamed) `spark.executor.processTreeMetrics.enabled`.

wypoon committed:
[SPARK-27324] Address feedback from Luca Canali and Imran Rashid.

Collect proc fs metrics if and only if spark.executor.processTreeMetrics.enabled
 is true and proc fs exists.

SparkQA commented:
**[Test build #117588 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117588/testReport)** for PR 27329 at commit [`f7ff9c2`](https://github.com/apache/spark/commit/f7ff9c2b3e071b4d04ce598acf7f47d831c419c2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22347/
Test PASSed.

rezasafi commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -233,6 +228,11 @@ package object config {
+  private[spark] val EXECUTOR_PROCESS_TREE_METRICS_ENABLED =
+    ConfigBuilder("spark.executor.processTreeMetrics.enabled")
+      .booleanConf
+      .createWithDefault(false)
```
shouldn't we add .doc for this metric and also spark.eventLog.logStageExecutorMetrics.enabled in addition to what is added to configuration.md?

wypoon replied:
Added.

rezasafi commented:
The goal of SPARK-24958, which introduces procfs metrics, was to compute peak procefs metrics along with other memory metrics and log them in event logs. So because of that there was a need that spark.eventLog.logStageExecutorMetrics.enabled being true when you want procfs metrics. 
I wasn't aware of SPARK-27189 when I objected the rename. After SPARK-27189  this has been changed. The metric system doesn't deal with peak values, but it exposed the processtree metrics. So the rename make sense now to me. 
Still I think we need to document somewhere that if you need peak metrics  for processtree metrics, you need to set both spark.eventLog.logStageExecutorMetrics.enabled and whatever name you decided for enabling for processtree metrics.

wypoon commented:
> Still I think we need to document somewhere that if you need peak metrics for processtree metrics, you need to set both spark.eventLog.logStageExecutorMetrics.enabled and whatever name you decided for enabling for processtree metrics.

I think this is clear from the documentation I added.

squito commented:
lgtm assuming tests pass & add the `.doc` as suggested by reza above.

rezasafi commented:
> > Still I think we need to document somewhere that if you need peak metrics for processtree metrics, you need to set both spark.eventLog.logStageExecutorMetrics.enabled and whatever name you decided for enabling for processtree metrics.
> 
> I think this is clear from the documentation I added.

yeah, I can see that. lgtm

wypoon committed:
[SPARK-27324] Add .doc to configurations.

SparkQA commented:
**[Test build #117598 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117598/testReport)** for PR 27329 at commit [`49a61db`](https://github.com/apache/spark/commit/49a61db1fbcd60cd3f3e0c83b8df958ea29960ef).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22357/
Test PASSed.

wypoon committed:
[SPARK-27324] Add back mistakenly deleted sentence from monitoring.md.

wypoon commented:
Argh, I misread the metrics (in `docs/monitoring.md`) @LucaCanali referred to in his first comment. I though he was referring to the Dropwizard metrics he worked on, but actually that section was added by Lantao Jin as documentation for what Edwina Lu added. So I mistakenly deleted a sentence from that section, which I have now restored.
Sorry for the confusion.

SparkQA commented:
**[Test build #117600 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117600/testReport)** for PR 27329 at commit [`9d5350a`](https://github.com/apache/spark/commit/9d5350a10df93f66ecdfeaff072e16fe40b29dc9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22359/
Test PASSed.

SparkQA commented:
**[Test build #117588 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117588/testReport)** for PR 27329 at commit [`f7ff9c2`](https://github.com/apache/spark/commit/f7ff9c2b3e071b4d04ce598acf7f47d831c419c2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117588/
Test PASSed.

SparkQA commented:
**[Test build #117598 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117598/testReport)** for PR 27329 at commit [`49a61db`](https://github.com/apache/spark/commit/49a61db1fbcd60cd3f3e0c83b8df958ea29960ef).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117598/
Test PASSed.

SparkQA commented:
**[Test build #117600 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117600/testReport)** for PR 27329 at commit [`9d5350a`](https://github.com/apache/spark/commit/9d5350a10df93f66ecdfeaff072e16fe40b29dc9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117600/
Test PASSed.

LucaCanali commented:
Thanks @wypoon for the changes, LGTM.
In case you are interested, I would like to propose a small change to monitoring.md that I think may be a good fit here as an addition/complement to the PR (or I can do a separate PR for that, if preferred). At line 643, Delete the current text and replace it with an updated one with more details:

Delete:
"""
Metrics `peakExecutorMetrics.*` are only enabled if `spark.eventLog.logStageExecutorMetrics.enabled` is true. 
"""

Replace with:
"""
Executor metric values and their measured peak values per executor are exposed via the REST API at the end point `/applications/[app-id]/executors`.
In addition, aggregated per-stage peak values of the executor metrics are written to the event log if `spark.eventLog.logStageExecutorMetrics.enabled` is true.
Executor metrics are also exposed via the Spark metrics system based on the Dropwizard metrics library.
"""

wypoon commented:
@LucaCanali thank you for the clarifying text. I shall do it.

wypoon committed:
[SPARK-27324] Use clarifying text in docs/monitoring.md from Luca Canali.

SparkQA commented:
**[Test build #117680 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117680/testReport)** for PR 27329 at commit [`5e134dd`](https://github.com/apache/spark/commit/5e134dd1a81e2b9ec20c3f3bfe2bb76172a4a639).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22442/
Test PASSed.

squito commented:
merged to master, thanks everyone!

SparkQA commented:
**[Test build #117680 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117680/testReport)** for PR 27329 at commit [`5e134dd`](https://github.com/apache/spark/commit/5e134dd1a81e2b9ec20c3f3bfe2bb76172a4a639).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117680/
Test FAILed.

wypoon commented:
Hmm, between the last passing test build (117600) and the next one that failed (117680), there was no code change, only a change in docs/monitoring.md. Thus the test failure cannot be caused by my final commit. Perhaps the test is flaky?
